packages/db-schema/src/indexing/defs.ts

```diff
@@ -0,0 +1,204 @@
+      t.blockTimestamp,
+      t.logIndex,
+    ),
+    // Filter by type over time:
+    typeBlocknumIdx: index("events_type_blocknum_idx").on(
+      t.type,
+      sql`${t.blockNumber} DESC`,
+    ),
+    blockNumberIdx: index("events_block_number_idx").on(t.blockNumber),
+    // Helpful for exact block lookups / reorg checks
+    blockHashIdx: index("events_block_hash_idx").on(t.blockHash),
+    // Housekeeping / retention jobs
+    createdAtIdx: index("events_created_at_idx").on(t.createdAt),
+  }),
+);
+
+/**
+ * Off-chain mirror of on-chain stake positions.
+ * One row per stake created on chain (keyed by the on-chain stake/receipt id, e.g. NFT token id).
+ * The row is updated as lifecycle transitions occur (stake → unstake → withdraw / relock).
+ *
+ * Contrast to `stake_changes`:
+ * - `stakes` is the *current mutable state* of each on-chain stake.
+ * - `stake_changes` is the *immutable ledger* of applied events.
+ * - Every mutation of `stakes` MUST be in the same DB transaction as inserting the corresponding `stake_changes` row.
+ *
+ * Invariants / notes:
+ * - `id` exactly matches the on-chain stake/receipt id (bigint).
+ * - `wallet` is a canonicalized (lowercased) EVM address.
+ * - `amount` is the *current active amount* for this stake (after partial unstakes/relocks).
+ * - Lifecycle columns (`stakedAt`, `canUnstakeAfter`, `unstakedAt`, `canWithdrawAfter`, `withdrawnAt`, `relockedAt`)
+ *   are progressively filled; `NULL` means “not reached yet”.
+ * - All timestamps are UTC (`timestamp` without time zone).
+ *
+ * Concurrency / idempotency:
+ * - Apply chain events exactly once using `(tx_hash, log_index)` uniqueness in `stake_changes`.
+ * - Use optimistic checks / row locks when updating `stakes` to avoid lost updates.
+ *
+ * Reorg handling:
+ * - Keep `block_number` + `block_hash` in `stake_changes` (and in the raw `indexing_events`)
+ *   to detect and repair on replays if a reorg invalidates prior assumptions.
+ * Lifecycle / status (derived from columns):
+ * - Locked:              unstakedAt IS NULL AND now() <  canUnstakeAfter
+ * - Unlocked:            unstakedAt IS NULL AND now() >= canUnstakeAfter
+ * - Unstaked:            unstakedAt IS NOT NULL AND withdrawnAt IS NULL
+ * - Withdrawal Pending:  withdrawnAt IS NULL AND canWithdrawAfter IS NOT NULL AND now() >= canWithdrawAfter
+ * - Withdrawn:           withdrawnAt IS NOT NULL            // terminal
+ * - Relocked:            relockedAt  IS NOT NULL            // terminal
+ * Notes:
+ * - “Withdrawn” and “Relocked” both end the stake lifecycle; they should be mutually
+ *   exclusive.
+ * - Status is time-relative (uses NOW at query time); if you need a historical view,
+ *   compute against a chosen reference timestamp instead of NOW().
+ */
+const stakes = pgTable(
+  "stakes",
+  {
+    id: bigint("id", { mode: "bigint" }).primaryKey().notNull(), // on-chain receipt / NFT id
+    wallet: bytea("wallet").notNull(), // canonicalized EVM addr
```

Comment:
I see here and elsewhere storing wallet address (and other values) as `bytea`. Is this for space savings or what? Also just wondering what the practical implications are... Will there be an easy way to browse the data and see human readable wallet addresses or write a SQL query and filter by wallet address?

Reply:
`SELECT * FROM table WHERE idemKey = '\xDEADBEEF'::bytea;` Like this.